README.md
# Filesize

[![Latest Version on Packagist](https://img.shields.io/packagist/v/resohead/filesize.svg?style=flat-square)](https://packagist.org/packages/resohead/filesize)
[![Total Downloads](https://img.shields.io/packagist/dt/resohead/filesize.svg?style=flat-square)](https://packagist.org/packages/resohead/filesize)

PHP fluent API extension for converting file sizes - like Carbon but for file sizes.

## Installation

You can install the package via composer:

```bash
composer require resohead/filesize
```

## Usage

### Basics

```php
$filesize = new Filesize;
```

```php
$filesize->from(1200)->toKilobytes()->asNumber(); // 1.0
$filesize->fromBytes(1200)->toKilobytes()->asNumber(2); // 1.2
$filesize->fromBytes(1200)->toKibibytes()->asNumber(2); // 1.17

$filesize->fromKilobytes(1)->toKibibytes()->round(3)->asNumber(); // 0.977
$filesize->fromKilobytes(1)->toBytes()->asInteger(); // 1000

$filesize->fromKibibytes(1)->toBytes()->asInteger(); // 1024

$filesize->fromBytes(1024)->toKilobytes()->round(3)->asNumber(); //1.024
$filesize->fromBytes(1000)->toKilobytes()->round(3)->asNumber(); //1.000

$filesize->fromBytes(1024)->toKibibytes()->round(3)->asNumber(); //1.000
$filesize->fromBytes(1000)->toKibibytes()->round(3)->asNumber(); //0.977

$filesize->fromKilobytes(1.2)->asNumber(); // 1200
$filesize->fromKilobytes(1.2)->toSame()->asNumber(); // 1.2

$filesize->fromGigabytes(5000))
        ->toTerabytes()
        ->asNumber(2); // 5.0

$filesize->fromGibibytes(5000))
        ->toTebibytes()
        ->asNumber(2); // 4.88

$filesize->fromGigabytes(1)
    ->toMegabytes()
    ->round(3)
    ->withThousandSeparator() // optional - default: ","
    ->withDecimalSeparator() // optional - default: "."
    ->asString(); // "1,073.742 MB"
```

### For Humans
This will automatically convert the given size to the most useful human format (defaults to decmial, i.e. GB)

```php

$filesize->fromBytes(1073741824)->round(2)->forHumans(); // "1.07 GB"

$filesize->fromGigabytes(0.5)->forHumans(); // "512 MB"

$filesize->fromBytes(15000)->inBinary()->round(1)->forHumans(); // '15.0 kB', or
$filesize->fromBytes(15000)->inBinary()->round(1)->forHumans(); // '14.6 KiB', or
$filesize->fromBytes(15000)->toKibibytes()->round(1)->forHumans(); // '14.6 KiB'

$filesize->fromBytes(15000)->inDecimal()->round(1)->forHumans(); // '15.0 kB', or
$filesize->fromBytes(15000)->toKibibytes()->round(1)->forHumans(); // '15.0 kB'


$filesize->fromBytes(1073741824)->inDecimal()->forHumans(); // "1.00 GiB"
$filesize->fromBytes(1073741824)->toKibtyes()->forHumans(); // "1.00 GiB"
$filesize->fromBytes(1073741824)->toMegabytes()->forHumans(); // "1.07 GB"
$filesize->fromBytes(1073741824)->inDecimal()->forHumans(); // "1.07 GB"

$filesize->fromMegabytes(56156113)->forHumans(); // "53.55 TB"
```

### Parse input from string
``` php
Filesize::parse('1 KB')->toBytes()->asInteger(); // 1024
Filesize::parse('1 KB')->toKilobytes()->asString(); // "1 kB"
Filesize::parse('1 KB')->round(3)->toMb()->asString(); // "0.001 MB"
```

### Testing

``` bash
composer test
```

## License

The MIT License (MIT). Please see [License File](LICENSE.md) for more information.


CHANGELOG.md
# Changelog

All notable changes to `filesize` will be documented in this file

## 1.0.0 - 201X-XX-XX

- initial release


tests/Unit.php
<?php

use Resohead\Filesize\Filesize;

it('it uses bytes as default input', function () {
    expect((new Filesize))
        ->from(1000)->asInteger()->toBe(1000)
        ->fromBytes(1000)->asInteger()->toBe(1000);
});

it('it returns bytes by default', function () {
    expect((new Filesize))
        ->fromKilobytes(1.5)->asInteger()->toBe(1500)
        ->fromKibibytes(1.5)->asInteger()->toBe(1536);
});

it('sets the base when outputting to formats', function(){
    expect((new Filesize)->fromKilobytes(1.5))
        ->byteUnitIndex(Filesize::BYTE)->toBe(0)
        ->byteUnitArray(Filesize::BYTE)->toBe(Filesize::BINARY_BYTE_UNITS)
        ->byteUnitIndex(Filesize::KILOBYTE)->toBe(1)
        ->byteUnitArray(Filesize::KILOBYTE)->toBe(Filesize::DECIMAL_BYTE_UNITS)
        ->byteUnitIndex(Filesize::KIBIBYTE)->toBe(1)
        ->byteUnitArray(Filesize::KIBIBYTE)->toBe(Filesize::BINARY_BYTE_UNITS);
});

it('can set the byte standard as string', function () {
    expect((new Filesize)->fromBytes(1610612736))
        ->toGibibytes()->asString()->toBe('1.50 GiB')
        ->toGigabytes()->asString()->toBe('1.61 GB');
});

it('can return the same unit', function(){

    expect((new Filesize)->fromKilobytes(1.60))
        ->asNumber()->toBe(1600.0);

    expect((new Filesize)->fromGigabytes(1.60))
        ->asNumber()->toBe(1600000000.0)
        ->toSame()->round(2)->asNumber()->toBe(1.60);

    expect((new Filesize)->fromGigabytes(1.61))
        ->asNumber()->toBe(1610000000.0)
        ->toGigabytes()->round(2)->asNumber()->toBe(1.61)
        ->toSame()->round(2)->asNumber()->toBe(1.61);
});

it('can round numbers', function(){
    expect((new Filesize)->fromBytes(1610000000))
        ->toMebibytes()->asNumber(2)->toBe(1535.42);
});

it('does something with base standard', function(){
    expect((new Filesize)->fromBytes(1610612736))
        ->toGigabytes()->asNumber()->toBe(1.61)
        ->inBinary()->asNumber()->toBe(1.50);

    expect((new Filesize)->fromGigabytes(1.61))
        ->toGigabytes()->inBinary()->asNumber()->toBe(1.50)
        ->toSame()->asNumber()->toBe(1.50)
        ->toSame()->inDecimal()->asNumber()->toBe(1.61)
        ->toSame()->inBinary()->asNumber()->toBe(1.50);
});

it('can set a default format for humans', function ($bytes, $decimal) {
    expect((new Filesize)->fromBytes($bytes))
        ->inDecimal()->forHumans()->toBe($decimal);
})->with([
    [1, '1 B'],
    [1000, '1 kB'],
    [2000, '2 kB'],
    [15000, '15 kB'],
    [1000000, '1.0 MB'],
    [1000000000, '1.00 GB'],
    [1500000000, '1.50 GB']
]);

it('can set binary format for humans', function ($bytes, $decimal, $decimalOneDp) {
    expect((new Filesize)->fromBytes($bytes))
        ->inBinary()->forHumans()->toBe($decimal);

        expect((new Filesize)->fromBytes($bytes))
            ->inBinary()->round(1)->forHumans()->toBe($decimalOneDp);
})->with([
    [1, '1 B', '1.0 B'],
    [1000, '1 KiB', '1.0 KiB'],
    [2000, '2 KiB', '2.0 KiB'],
    [15000, '15 KiB', '14.6 KiB'],
    [1000000, '1.0 MiB', '1.0 MiB'],
    [1000000000, '0.93 GiB', '0.9 GiB'],
    [1500000000, '1.40 GiB','1.4 GiB'],
]);

it('returns the decimal byte standard by default for humans', function () {
    expect((new Filesize)->fromGibibytes(1.5))
        ->forHumans()->toBe('1.61 GB');

    expect((new Filesize)->fromGibibytes(1.5))
        ->inDecimal()->forHumans()->toBe('1.61 GB');

    expect((new Filesize)->fromGigabytes(1.5))
        ->forHumans()->toBe('1.50 GB');

    expect((new Filesize)->fromGigabytes(1.5))
        ->inBinary()->forHumans()->toBe('1.40 GiB');
});

it('can set the byte standard for humans', function () {
    expect((new Filesize)->fromKilobytes(1610000))
        ->toGigabytes()->asNumber()->toBe(1.61)
        ->toGibibytes()->asNumber()->toBe(1.5)
        ->toGigabytes()->forHumans()->toBe('1.61 GB')
        ->toGibibytes()->forHumans()->toBe('1.50 GiB');
});

it('can format for humans', function(){
    expect((new Filesize)->fromBytes(1610612736))
        ->toGibibytes()->round(2)->asNumber()->toBe(1.50)
        ->toGibibytes()->forHumans()->toBe('1.50 GiB')
        ->toGibibytes()->round(1)->forHumans()->toBe('1.5 GiB')
        ->toGibibytes()->round(2)->forHumans()->toBe('1.50 GiB')
        ->toGigabytes()->round(2)->asNumber()->toBe(1.61);
});

it('can format as a string', function () {
    expect((new Filesize)->fromBytes(1024))
        ->asString()->toBe('1,024 B')
        ->withThousandSeparator()->asString()->toBe('1,024 B')
        ->withThousandSeparator('.')->asString()->toBe('1.024 B')
        ->withoutThousandSeparator()->asString()->toBe('1024 B')
        ->withDecimalSeparator(',')->asString()->toBe('1024 B')
        ->withThousandSeparator('.')->withDecimalSeparator(',')->asString()->toBe('1.024 B');
});

it('can round numbers formatted as string', function () {
    expect((new Filesize)->fromKilobytes(1.5)->toKilobytes())
        ->asString()->toBe('2 kB')
        ->round(0)->asString()->toBe('2 kB')
        ->round(1)->asString()->toBe('1.5 kB')
        ->round(2)->asString()->toBe('1.50 kB')
        ->round(2)->withDecimalSeparator(',')->asString()->toBe('1,50 kB');

    expect((new Filesize)->fromKibibytes(1.5)->toKilobytes())
        ->asString()->toBe('2 kB')
        ->round(0)->asString()->toBe('2 kB')
        ->round(1)->asString()->toBe('1.5 kB')
        ->round(2)->asString()->toBe('1.54 kB')
        ->round(2)->withDecimalSeparator(',')->asString()->toBe('1,54 kB');
});

it('it uses the latest byte standard for calculations', function(){
    expect((new Filesize)->fromKilobytes(1.5))
        ->toKibibytes()->asNumber(2)->toBe(1.46)
        ->toKibibytes()->inDecimal()->asNumber(2)->toBe(1.5)
        ->toKibibytes()->inBinary()->asNumber(2)->toBe(1.46)
        ->toKilobytes()->inBinary()->asNumber(2)->toBe(1.46);
});

it('can parse file sizes', function ($input, $output) {
    expect(Filesize::parse($input))
        ->toBytes()->asNumber()->toBe($output);
})->with([
    ['1', 1.0],
    ['12', 12.0],
    ['12 B', 12.0],
    ['12 B', 12.0],
    ['12.0 B', 12.0],
    ['12.0B', 12.0],
    ['12.0b', 12.0],
    ['12.0 b', 12.0],
    ['1 kb', 1000.0],
    ['1 kB', 1000.0],
    ['1 kib', 1024.0],
    ['1 KiB', 1024.0],
    ['1 KIB', 1024.0],
    ['1MB', 1000000.0],
    ['1 MB', 1000000.0],
    ['1.0 MB', 1000000.0],
    ['1.0MB', 1000000.0],
    ['1.0mb', 1000000.0],
    ['1.0 mb', 1000000.0],
    ['1.0 mib', 1048576.0],
    ['1.0 MIB', 1048576.0],
    ['1.0 MiB', 1048576.0]
]);

it('can parse large file sizes', function () {
    expect(Filesize::parse('1073741824 GB'))
        ->toGib()->asInteger()->toBe(1000000000)
        ->toGb()->asInteger()->toBe(1073741824);
});

test('it can convert from bytes to kb', function () {
    expect((new Filesize)->fromBytes(1_000))
        ->toKb()->asInteger()->toEqual(1);
});

test('it can convert from bytes to kib', function () {
    expect((new Filesize)->fromBytes(1_024))
        ->toKiB()->asNumber()->toEqual(1.0)
        ->toKb()->round(3)->asNumber()->toEqual(1.024);
});

test('it can convert from bytes to mb', function () {
    expect((new Filesize)->fromBytes(1_000_000))
        ->toMb()->asInteger()->toEqual(1);
});

test('it can convert from bytes to mib', function () {
    expect((new Filesize)->fromBytes(1_048_576))
        ->toMiB()->asNumber()->toEqual(1.0);
});

test('it can convert from bytes to gb', function () {
    expect((new Filesize)->fromBytes(1_000_000_000))
        ->toGb()->asInteger()->toEqual(1)
        ->toGib()->round(2)->asNumber(0.93);
});

test('it can convert from bytes to gib', function () {
    expect((new Filesize)->fromBytes(1_073_741_824))
        ->toMebibytes()->asNumber()->toEqual(1024)
        ->toKibibytes()->asNumber()->toEqual(1_048_576)
        ->toKilobytes()->asNumber(2)->toEqual(1_073_741.82)
        ->toGiB()->round(2)->asNumber()->toEqual(1.00)
        ->toGb()->round(2)->asNumber()->toEqual(1.07);
});

it('converts kilobytes', function(){
    expect(new Filesize)
        ->fromKilobytes(1)->toKibibytes()->round(3)->asNumber()->toBe(0.977)
        ->fromKilobytes(1)->toBytes()->asInteger()->toBe(1000)
        ->fromKibibytes(1)->toBytes()->asInteger()->toBe(1024)
        ->fromBytes(1024)->toKilobytes()->round(3)->asNumber()->toBe(1.024)
        ->fromBytes(1000)->toKilobytes()->round(3)->asNumber()->toBe(1.000)
        ->fromBytes(1024)->toKibibytes()->round(3)->asNumber()->toBe(1.000)
        ->fromBytes(1000)->toKibibytes()->round(3)->asNumber()->toBe(0.977);
});

it('converts megabytes', function(){
    expect(new Filesize)
        ->fromMegabytes(1)->toMebibytes()->round(2)->asNumber()->toBe(0.95)
        ->fromMegabytes(1)->toKilobytes()->asInteger()->toBe(1000)
        ->fromMegabytes(1)->toBytes()->asInteger()->toBe(1_000_000)
        ->fromMebibytes(1)->toKibibytes()->asInteger()->toBe(1024)
        ->fromMebibytes(1)->toKilobytes()->round(2)->asNumber()->toBe(1048.58)
        ->fromMebibytes(1)->toBytes()->asInteger()->toBe(1_048_576);
});


src/Filesize.php
<?php

namespace Resohead\Filesize;

use Exception;

class Filesize
{
    protected $bytes;
    protected $precision = null;
    protected string $decimalSeparator = '.';
    protected string $thousandSeparator = ',';
    protected ?string $unit = null;
    protected string $from = self::BYTE;
    protected int $base = 10;

    const BYTE = 'B';
    const KILOBYTE = 'kB';
    const MEGABYTE = 'MB';
    const GIGABYTE = 'GB';
    const TERABYTE = 'TB';
    const PETABYTE = 'PB';
    const EXABYTE = 'EB';
    const ZETTABYTE = 'ZB';
    const YOTTABYTES = 'YB';

    const KIBIBYTE = 'KiB';
    const MEBIBYTE = 'MiB';
    const GIBIBYTE = 'GiB';
    const TEBIBYTE = 'TiB';
    const PEBIBYTE = 'PiB';
    const EXBIBYTE = 'EiB';
    const ZEBIBYTE = 'ZiB';
    const YOBIBYTE = 'YiB';

    const DECIMAL_BYTE_UNITS = [
        self::BYTE,
        self::KILOBYTE,
        self::MEGABYTE,
        self::GIGABYTE,
        self::TERABYTE,
        self::PETABYTE,
        self::EXABYTE,
        self::ZETTABYTE,
        self::YOTTABYTES,
    ];

    const BINARY_BYTE_UNITS = [
        self::BYTE,
        self::KIBIBYTE,
        self::MEBIBYTE,
        self::GIBIBYTE,
        self::TEBIBYTE,
        self::PEBIBYTE,
        self::EXBIBYTE,
        self::ZEBIBYTE,
        self::YOBIBYTE,
    ];

    const BYTE_PRECISION = [0, 0, 1, 2, 2, 3, 3, 4, 4];

    const BYTE_BINARY = 1024;
    const BYTE_DECIMAL = 1000;

    public function __toString()
    {
        return (string) $this->forHumans();
    }

    public function getDefaultUnit()
    {
        return $this->unit ?? $this->from;
    }

    /**
     * Convert bytes to human readable format based on size.
     *
     * @return string Human readable bytes
     */
    public function forHumans(): string
    {
        $bytes = $this->bytes;
        $array = self::DECIMAL_BYTE_UNITS;
        $byteStandard = self::BYTE_DECIMAL;

        if($this->isBinary()){
            $array = self::BINARY_BYTE_UNITS;
            $byteStandard = self::BYTE_BINARY;
        }

        for ($i = 0; ($bytes / $byteStandard) >= 0.9 && $i < count($array)-1; $i++) {
            $bytes /= $byteStandard;
        }
        return number_format($bytes, $this->precision ??= self::BYTE_PRECISION[$i]) . ' '. $array[$i];
    }

    /**
     * Convert bytes to fixed human readable format
     *
     * @return string Human readable bytes
     */
    public function asString(): string
    {
        return number_format($this->convert(), $this->getPrecision(), $this->decimalSeparator, $this->thousandSeparator) . " " . $this->getDefaultUnit();
    }

    protected function getPrecision(): string
    {
        return $this->precision ??= self::BYTE_PRECISION[$this->byteUnitIndex($this->unit)];
    }

    public function asNumber($precision = null): int|float
    {
        return round($this->convert(), $precision ?? $this->getPrecision());
    }

    public function asInteger(): int
    {
        return (int) round($this->convert());
    }

    public function round($precision = null): self
    {
        $this->precision = $precision;
        return $this;
    }

    public function getByteStandard(): int
    {
        return $this->isDecimal() ? self::BYTE_DECIMAL : self::BYTE_BINARY;
    }

    public function getByteStandardFromUnit($unit): int
    {
        if($unit === self::BYTE){
            return $this->getByteStandard();
        }

        if(array_search($unit, self::BINARY_BYTE_UNITS)){
            return self::BYTE_BINARY;
        }

        return self::BYTE_DECIMAL;
    }

    protected function getBaseFromUnit($unit)
    {
        return $this->isBinaryByte($unit)
            ? 2
            : 10;
    }

    protected function isBinaryByte($unit)
    {
        return in_array($unit, self::BINARY_BYTE_UNITS);
    }

    public function byteUnitArray($unit)
    {
        return $this->isBinaryByte($unit)
            ? self::BINARY_BYTE_UNITS
            : self::DECIMAL_BYTE_UNITS;
    }

    public function byteUnitIndex($unit)
    {
        return array_search($unit, $this->byteUnitArray($unit));
    }

    public function multiplier($unitsTo)
    {
        return $this->getByteStandardFromUnit($unitsTo) ** $this->byteUnitIndex($unitsTo);
    }

    protected function convert(): int|float
    {
        return $this->bytes / $this->multiplier($this->unit);
    }

    protected function convertToBytes(int|float $size = 0, string $unit = self::BYTE)
    {
        return $size * $this->multiplier($unit);
    }

    public function base(int $base): self
    {
        if($base !== 2 && $base !== 10){
            throw new Exception('Invalid base value. Must be 2 or 10');
        }
        $this->base = $base;
        return $this;
    }

    public function inDecimal(): self
    {
        // map the current output index with the decimal index
        $this->unit = self::DECIMAL_BYTE_UNITS[$this->byteUnitIndex($this->getDefaultUnit())];
        return $this->base(10);
    }

    public function inBinary(): self
    {
        // map the current output index with the binary index
        $this->unit = self::BINARY_BYTE_UNITS[$this->byteUnitIndex($this->getDefaultUnit())];
        return $this->base(2);
    }

    protected function isDecimal()
    {
        return $this->base === 10;
    }

    protected function isBinary()
    {
        return $this->base === 2;
    }

    public function from(int|float $size = 0, string $unit = self::BYTE): self
    {
        $this->bytes = $this->convertToBytes($size, $unit);
        $this->from = $unit;
        // $this->unit = $unit;
        return $this;
    }

    public function fromBytes($size = 0): self
    {
        return $this->from($size, self::BYTE);
    }

    public function fromKilobytes($size = 0): self
    {
        return $this->from($size, self::KILOBYTE);
    }

    public function fromKibibytes($size = 0): self
    {
        return $this->from($size, self::KIBIBYTE);
    }

    public function fromMegabytes($size = 0): self
    {
        return $this->from($size, self::MEGABYTE);
    }

    public function fromMebibytes($size = 0): self
    {
        return $this->from($size, self::MEBIBYTE);
    }

    public function fromGigabytes($size = 0): self
    {
        return $this->from($size, self::GIGABYTE);
    }

    public function fromGibibytes($size = 0): self
    {
        return $this->from($size, self::GIBIBYTE);
    }

    public function fromTerabytes($size = 0): self
    {
        return $this->from($size, self::TERABYTE);
    }

    public function fromPetabytes($size = 0): self
    {
        return $this->from($size, self::PETABYTE);
    }

    public function fromExabytes($size = 0): self
    {
        return $this->from($size, self::EXABYTE);
    }

    public function fromZettabytes($size = 0): self
    {
        return $this->from($size, self::ZETTABYTE);
    }

    public function fromYottabytes($size = 0): self
    {
        return $this->from($size, self::YOTTABYTES);
    }

    public static function parse($input)
    {
        preg_match('/(?P<size>-?[0-9\.]+)\s*(?P<unit>[A-z]+b|[bB])?$/i', trim($input), $matches, PREG_OFFSET_CAPTURE);

        $matchedUnit = strtoupper($matches['unit'][0] ?? 'B');
        $decimalBytes = array_search($matchedUnit, array_map('strtoupper', self::DECIMAL_BYTE_UNITS));
        $binaryBytes = array_search($matchedUnit, array_map('strtoupper', self::BINARY_BYTE_UNITS));

        $unit = null;
        if($decimalBytes !== false){
            $unit = self::DECIMAL_BYTE_UNITS[$decimalBytes];
        } elseif($binaryBytes !== false) {
            $unit = self::BINARY_BYTE_UNITS[$binaryBytes];
        }

        if (!isset($matches['size']) || !isset($unit)) {
            throw new Exception("Could not parse \"{$input}\"");
        }

        return (new self)->from((float) $matches['size'][0], $unit);
    }

    public function to($unit): self
    {
        $this->unit = $unit;
        $this->base(
            $this->getBaseFromUnit($unit)
        );
        return $this;
    }

    public function toSame()
    {
        $this->unit = $this->getDefaultUnit();
        return $this;
    }

    public function toBytes(): self
    {
        return $this->to(self::BYTE);
    }

    public function toB(): self
    {
        return $this->toBytes();
    }

    public function toKilobytes(): self
    {
        return $this->to(self::KILOBYTE);
    }

    public function toKibibytes(): self
    {
        return $this->to(self::KIBIBYTE);
    }

    public function toKb(): self
    {
        return $this->toKilobytes();
    }

    public function toKiB(): self
    {
        return $this->toKibibytes();
    }

    public function toMegabytes(): self
    {
        return $this->to(self::MEGABYTE);
    }

    public function toMebibytes(): self
    {
        return $this->to(self::MEBIBYTE);
    }

    public function toMb(): self
    {
        return $this->toMegabytes();
    }

    public function toMib(): self
    {
        return $this->toMebibytes();
    }

    public function toGigabytes(): self
    {
        return $this->to(self::GIGABYTE);
    }

    public function toGibibytes(): self
    {
        return $this->to(self::GIBIBYTE);
    }

    public function toGb(): self
    {
        return $this->toGigabytes();
    }

    public function toGib(): self
    {
        return $this->toGibibytes();
    }

    public function toTerabytes(): self
    {
        return $this->to(self::TERABYTE);
    }

    public function toTb(): self
    {
        return $this->toTerabytes();
    }

    public function toTebibytes(): self
    {
        return $this->to(self::TEBIBYTE);
    }

    public function toTib(): self
    {
        return $this->toTebibytes();
    }

    public function toPetabytes(): self
    {
        return $this->to(self::PETABYTE);
    }

    public function toPebibytes(): self
    {
        return $this->to(self::PEBIBYTE);
    }

    public function toPb(): self
    {
        return $this->toPetabytes();
    }

    public function toPib(): self
    {
        return $this->toPebibytes();
    }

    public function toExabytes(): self
    {
        return $this->to(self::EXABYTE);
    }

    public function toExbibyte(): self
    {
        return $this->to(self::EXBIBYTE);
    }

    public function toEb(): self
    {
        return $this->toExabytes();
    }

    public function toEib(): self
    {
        return $this->toExbibyte();
    }

    public function toZettabytes(): self
    {
        return $this->to(self::ZETTABYTE);
    }

    public function toZebibytes(): self
    {
        return $this->to(self::ZEBIBYTE);
    }

    public function toZb(): self
    {
        return $this->toZettabytes();
    }

    public function toZib(): self
    {
        return $this->toZebibytes();
    }

    public function toYottabytes(): self
    {
        return $this->to(self::YOTTABYTES);
    }

    public function toYobibytes(): self
    {
        return $this->to(self::YOBIBYTE);
    }

    public function toYb(): self
    {
        return $this->toYottabytes();
    }

    public function toYib(): self
    {
        return $this->toYobibytes();
    }

    public function withDecimalSeparator(string $separator): self
    {
        $this->decimalSeparator = $separator;
        return $this;
    }

    public function withThousandSeparator(string $separator = ','): self
    {
        $this->thousandSeparator = $separator;
        return $this;
    }

    public function withoutThousandSeparator(): self
    {
        $this->thousandSeparator = '';
        return $this;
    }
}
